sherpa/models/model.py

```diff
@@ -171,72 +170,44 @@ def __iter__(self):
                 'use {} instead'.format(nname)
             warnings.warn(wmsg, DeprecationWarning)
 
-        # If .pars is set up, use this as it is a micro-optimisation,
-        # otherwise (e.g. during calls to __init__), fall back to
-        # scanning through the dictionary for any code called during
-        # model initialisation (alternatively it could be considered
-        # to be an error to use the wrong capitalisation during the
-        # construction phase, but this would be a policy change so is
-        # not enforced here).
-        #
-        # How much of this should be cached?
-        #
-        pars = self.__dict__.get('pars')
-        if pars is not None:
-            for par in pars:
-                if par.name.lower() == lname:
-                    return par
-
-            rpars = self.__dict__.get('_renamedpars')
-            if rpars is not None:
-                for (oname, nname) in rpars:
-                    if oname.lower() == lname:
-                        for par in pars:
-                            if par.name == nname:
-                                warn(oname, nname)
-                                return par
+        parameter = self._par_index.get(lowered_name)
 
-        else:
-            for key in self.__dict__:
-                try:
-                    kname = key.lower()
-                except AttributeError:
-                    continue
-
-                if lname == kname:
-                    val = self.__dict__.get(key)
-                    if isinstance(val, Parameter):
-                        return val
-
-            rpars = self.__dict__.get('_renamedpars')
-            if rpars is not None:
-                for (oname, nname) in rpars:
-                    if oname.lower() == lname:
-                        # Easier to call itself with the proper name
-                        # than to loop through __dict__ again. This
-                        # should not be a common case.
-                        ans = getattr(self, nname)
-                        warn(oname, nname)
-                        return ans
-
-        # this must be AttributeError for 'getattr' to work
-        raise AttributeError("'%s' object has no attribute '%s'" %
-                             (type(self).__name__, name))
+        if parameter is not None:
+            if lowered_name in parameter.aliases:
+                warn(lowered_name, parameter.name)
+            return parameter
+
+        NoNewAttributesAfterInit.__getattribute__(self, name)
 
     def __setattr__(self, name, val):
         par = getattr(self, name.lower(), None)
         if (par is not None) and isinstance(par, Parameter):
+            # When setting an attribute that is a Parameter, set the parameter's
+            # value instead.
             par.val = val
         else:
             NoNewAttributesAfterInit.__setattr__(self, name, val)
+            if isinstance(val, Parameter):
```

Comment:
Should we add a check here to ensure that the attribute name has the same case as the parameter name? As I mentioned in the original PR, I'm not sure that this is an intended constraint, but if it is there should be some enforcement/checks.

Or doesissue no-longer matter with the new design?

Reply:
Are you asking if we should enforce that the parameter name and the model's attribute it is assigned to have the same case? I wouldn't think that is a valid constraint. I am not really sure we should enforce them to be the same either, to be honest, but maybe we do? If we don't, does that mean we should use the parameter name (which might be different than the attribute name) as another alias for the model to respond to? We don't check that there aren't parameters overwriting other parameters either. Do we care?

Reply:
Given that I found an issue with my implementation when this didn't hold, perhaps all I need to do is convince myself it's not a problem with the new design and then leave it at that?

Reply:
Hmm... Am I missing at least one test case?

Reply:
Note that this is now all fuzzy in my memory, so maybe I'm making a mountain out of a non-existant molehill, but I'm thinking it's related to our discussion at https://github.com/sherpa/sherpa/pull/255#discussion_r135114243

Reply:
I see. I'll edit the relevant test to see if we are missing something.

Reply:
Doing so actually uncovered a bug both in the code and in the test.

Reply:
Yay for testing the testing.

Reply:
Say I have a model that defines a parameter with name `Ampl` and it assigns it to `self.ampl`, with alias `NORM`. The client calls `model.norm`. What should the warning say? What's the "right" name of the parameter? The current code would say "Ampl", but maybe it should be "ampl", shouldn't it? The problem is that there are two names, one is the name of the parameter, one is the name of the model attribute.

Reply:
Well, this is partly why I was talking about enforcing a name check at object creation time :-)

However, since the case of the name is "not important" - in that the case is ignored when using the model attribute to access the parameter name - I would use the case of the parameter name in any error message. That is stick with `Ampl` in the warning.

I doubt many users are going to see this warning (as I think the default is for Python to ignore these warnings), at least from the various REPLs.

Reply:
That's good enough for me, and what the code I was about to push does. In the end, I just had to edit the tests so they covered this case and compensate for the different expected results.